README.md
# docker: tor socks & dns proxy 🌐 🐳

Docker Hub: https://hub.docker.com/r/fphammerle/tor-proxy

Signed image tags: https://github.com/fphammerle/docker-tor-proxy/tags

```sh
$ sudo docker run --rm --name tor_proxy \
    -p 127.0.0.1:9050:9050/tcp \
    -p 127.0.0.1:53:9053/udp \
    fphammerle/tor-proxy
```

or after cloning the repository 🐙
```sh
$ sudo docker-compose up
```

### Test Proxies

```sh
$ curl --proxy socks5h://localhost:9050 ipinfo.io
$ torsocks wget -O - ipinfo.io
$ torsocks lynx -dump https://check.torproject.org/
$ dig @localhost fabian.hammerle.me
$ ssh -o 'ProxyCommand nc -x localhost:9050 -v %h %p' abcdefghi.onion
# no anonymity!
$ chromium-browser --proxy-server=socks5://localhost:9050 ipinfo.io
```

### Read-only Root Filesystem

Optionally add
```sh
$ sudo docker run --read-only -v tor_proxy_data:/var/lib/tor --tmpfs /tmp:rw,size=4k` …
```
to make the container's root filesystem read-only.

### Isolate Host

```sh
sudo iptables -A OUTPUT ! -o lo -j REJECT --reject-with icmp-admin-prohibited
```

### Change `SocksTimeout` Option

```sh
$ sudo docker run -e SOCKS_TIMEOUT_SECONDS=60 …
```

### Select Exit Nodes

```sh
$ sudo docker run -e EXIT_NODES=1.2.3.4,1.2.3.5,{at} …
```

### Exclude Exit Nodes

```sh
$ sudo docker run -e EXCLUDE_EXIT_NODES=1.2.3.4,1.2.3.5 …
```

### Show Circuits

```sh
$ printf 'AUTHENTICATE\nGETINFO circuit-status\nQUIT\n' \
    | sudo docker exec -i tor_proxy nc localhost 9051
```
relay search: https://metrics.torproject.org/rs.html

or using [onioncircuits](https://gitlab.tails.boum.org/tails/onioncircuits) ([debian repo](https://salsa.debian.org/pkg-privacy-team/onioncircuits)):
```sh
$ sudo apt-get install --no-install-recommends onioncircuits
$ sudo nsenter --target "$(sudo docker inspect --format='{{.State.Pid}}' tor_proxy)" --net \
    sudo -u $USER onioncircuits
```

### Troubleshooting

Disable `SafeLogging` (temporarily) to log IP addresses instead of `[scrubbed]`:
```
$ printf 'AUTHENTICATE\nSETCONF SafeLogging=0\nGETCONF SafeLogging\nQUIT\n' \
    | sudo docker exec -i tor_proxy nc localhost 9051
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- support for enabling `SafeSocks` option by assigning `1` to environment
  variable `SAFE_SOCKS` (instructs tor to block socks connections when provided
  with remote ip address instead of domain name to detect dns leaks due to
  applications resolving domain names without tor)

## [4.5.1] - 2022-02-27
### Fixed
- "sed: write error" for large values of `EXIT_NODES` and `EXCLUDE_EXIT_NODES`

## [4.5.0] - 2021-09-06
### Added
- added ability to select exit nodes explicitly via `EXIT_NODES` environment variable

## [4.4.0] - 2021-07-07
### Added
- added ability to exclude exit nodes via `EXCLUDE_EXIT_NODES` environment variable

## [4.3.1] - 2021-05-15
### Fixed
- create prerouting chain if not already created by container runtime
  (fixes dns proxy with direct routing & transparent proxy on some hosts)

## [4.3.0] - 2021-05-07
### Added
- when running entrypoint as `uid=0`:
  add `nftables` rule forwarding udp dns requests (port 53) to tor
  (alternative for less flexible `docker run --publish 53:9053 …`)

## [4.2.0] - 2021-04-30
### Added
- when running entrypoint as `uid=0`:
  add `nftables` rule redirecting `tcp` traffic to transparent proxy

## [4.1.0] - 2021-03-03
### Added
- enabled transparent proxy, listening on port `9040` (requires netfilter rules)
- image labels:
  - `org.opencontainers.image.revision` (git commit hash via build arg)
  - `org.opencontainers.image.source` (repo url)
  - `org.opencontainers.image.title`

## [4.0.0] - 2021-01-01
### Changed
- healthcheck: replace periodic http requests with probing
  [network-liveness](https://gitweb.torproject.org/torspec.git/tree/control-spec.txt)
  to improve privacy (keeping dns requests for faster updates)
- changed log level of `control` domain to `warn`
  (to avoid log spam by healthcheck connecting to control listener)
- added message domains to log messages

### Removed
- `curl`
- `bind-tools` package including `dig`

## [3.0.0] - 2020-10-03
### Added
- create mount point at `/var/lib/tor`
  to be able to make container's root filesystem read-only

### Changed
- moved tor's data directory from `/home/onion/.tor` to `/var/lib/tor`
- run `tor` as user `tor` (uid=100) instead of `onion` (uid=101)
- docker-compose & ansible-playbook: read-only root filesystem

### Fixed
- ansible-playbook: drop capabilities

## [2.1.0] - 2020-09-27
### Added
- enable [tor control](https://gitweb.torproject.org/torspec.git/tree/control-spec.txt)
  listener on port `9051`
  (listening on loopback device only)

## [2.0.0] - 2020-03-24
### Changed
- changed DNS port from 53 to 9053
- run as unprivileged user

## [1.1.1] - 2020-03-21
### Fixed
- fix invalid torrc path

## [1.1.0] - 2020-03-21
### Added
- change `SocksTimeout` option by setting `$SOCKS_TIMEOUT_SECONDS`

## [1.0.0] - 2019-10-12
### Added
- tor socks5 & DNS proxy

[Unreleased]: https://github.com/fphammerle/docker-tor-proxy/compare/v4.5.1...HEAD
[4.5.1]: https://github.com/fphammerle/docker-tor-proxy/compare/v4.5.0...v4.5.1
[4.5.0]: https://github.com/fphammerle/docker-tor-proxy/compare/v4.4.0...v4.5.0
[4.4.0]: https://github.com/fphammerle/docker-tor-proxy/compare/v4.3.1...v4.4.0
[4.3.1]: https://github.com/fphammerle/docker-tor-proxy/compare/v4.3.0...v4.3.1
[4.3.0]: https://github.com/fphammerle/docker-tor-proxy/compare/v4.2.0...v4.3.0
[4.2.0]: https://github.com/fphammerle/docker-tor-proxy/compare/v4.1.0...v4.2.0
[4.1.0]: https://github.com/fphammerle/docker-tor-proxy/compare/v4.0.0...v4.1.0
[4.0.0]: https://github.com/fphammerle/docker-tor-proxy/compare/v3.0.0...v4.0.0
[3.0.0]: https://github.com/fphammerle/docker-tor-proxy/compare/v2.1.0...v3.0.0
[2.1.0]: https://github.com/fphammerle/docker-tor-proxy/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/fphammerle/docker-tor-proxy/compare/v1.1.1...v2.0.0
[1.1.1]: https://github.com/fphammerle/docker-tor-proxy/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/fphammerle/docker-tor-proxy/compare/1.0.0...v1.1.0
[1.0.0]: https://github.com/fphammerle/docker-tor-proxy/releases/tag/1.0.0


entrypoint.sh
#!/bin/sh

set -e

if [ "$(id -u)" -eq 0 ]; then
    nft add chain ip nat PREROUTING { type nat hook prerouting priority dstnat \; } || true
    nft add rule ip nat PREROUTING ip protocol tcp fib daddr type != local counter redirect to :9040 \
        || echo 'warning: failed to configure nftables for transparent proxy (missing CAP_NET_ADMIN?)'
    nft add rule ip nat PREROUTING fib daddr type local udp dport 53 counter redirect to :9053 \
        || echo 'warning: failed to configure nftables for DNS proxy' \
                '(alternative for less flexible `docker run --publish 53:9053 ...`)'
    exec su -s /bin/sh tor -- "$0" "$@"
fi

sed -e "s/{safe_socks}/$SAFE_SOCKS/" /torrc.template > /tmp/torrc
# default: 120 sec
# https://github.com/torproject/tor/blob/tor-0.4.1.7/src/core/or/connection_edge.c#L1099
if [ -z "$SOCKS_TIMEOUT_SECONDS" ]; then
    sed -ie '/{socks_timeout_seconds}/d' /tmp/torrc
else
    sed -ie "s/{socks_timeout_seconds}/$SOCKS_TIMEOUT_SECONDS/" /tmp/torrc
fi

# > list of identity fingerprints, country codes, and address patterns
if [ ! -z "$EXIT_NODES" ]; then
    echo -n 'ExitNodes ' >> /tmp/torrc
    printenv EXIT_NODES >> /tmp/torrc
fi
if [ ! -z "$EXCLUDE_EXIT_NODES" ]; then
    echo -n 'ExcludeExitNodes ' >> /tmp/torrc
    printenv EXCLUDE_EXIT_NODES >> /tmp/torrc
fi

exec "$@"
